playwright-e2e/pages/page-base.ts

```diff
@@ -147,6 +151,13 @@ export default abstract class PageBase implements PageInterface {
     await new Checkbox(this.page, { label: 'I have already read and signed the informed consent document' }).check();
   }
 
+  async finish(): Promise<void> {
+    await Promise.all([
+      waitForNoSpinner(this.page),
+      this.getFinishButton().click()
```

Comment:
Great question! I'll do my best to explain.

The order of the promises in `Promise.all` is maintained. The first promise in the array will get resolved first.

More on my thinking about synchronization...

`waitForNoSpinner()` works by checking page one time. It doesn't wait and polls for few seconds to recheck for spinner. That would increase tests' run time.

(1) To check for spinner and then click, I can do this,
```
await waitForNoSpinner(this.page);
await this.getFinishButton().click();
```

(2) If I want to check spinner after click, I can do this,
```
await this.getFinishButton().click();
await waitForNoSpinner(this.page);
```

(3) If I want to check spinner before and after click, I'd have to do this,
```
await waitForNoSpinner(this.page);
await this.getFinishButton().click();
await waitForNoSpinner(this.page);
```
(4) Lastly, `Promise.all`. It works like the first approach.

None actually solves the sync issue 100% of time due to the fact Playwright runs really really fast, `waitForNoSpinner()` finishes in milliseconds if no spinner found. Hoever, spinner can be rendered after the check has finished.
Any one of four approaches will help to reduce the probability of it happens.

Reply:
`concurrent` is a misleading concept from `Promise.all` because JS code in a single-threaded language.

Reply:
Right.  I'm using concurrent in the context of the MDN documentation here:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#promise_concurrency:~:text=Note%20that%20JavaScript%20is%20single%2Dthreaded%20by%20nature%2C%20so%20at%20a%20given%20instant%2C%20only%20one%20task%20will%20be%20executing%2C%20although%20control%20can%20shift%20between%20different%20promises%2C%20making%20execution%20of%20the%20promises%20appear%20concurrent.

Yes - without workers, the language is single threaded.  I'm not meaning to imply otherwise.

Turning back to the logic flow, do you need to wait for the spinner is gone before clicking finish?  If so, I'd implement the logic in 2 or 3 above.